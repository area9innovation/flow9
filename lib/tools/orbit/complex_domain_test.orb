// Test of domain pattern matching with complex domain expressions

// Define quote
fn quote(e : ast) = e;

// Create a test graph
let g = makeOGraph("complex_domain_test");

// Create domain expressions and a value to annotate
let a_id = addOGraph(g, quote(a));
let b_id = addOGraph(g, quote(b));
let expr_id = addOGraph(g, quote(a + b));

// Add a simple domain
let algebra_id = addOGraph(g, quote(Algebra));

// Add a complex domain (function call)
let typefn_id = addOGraph(g, quote(Type("number")));

// Add a nested domain (a + b could be an expression or a domain)
let nested_id = addOGraph(g, quote(Group(a + b)));

// Print info about initial graph state
println("Created test graph with nodes:");
println("a with ID: " + i2s(a_id));
println("b with ID: " + i2s(b_id));
println("a + b with ID: " + i2s(expr_id));
println("Algebra with ID: " + i2s(algebra_id));
println("Type(\"number\") with ID: " + i2s(typefn_id));
println("Group(a + b) with ID: " + i2s(nested_id));

// Add domains to the expression
println("\nAdding domains to a + b");
let add_algebra = addDomainToNode(g, expr_id, algebra_id);
println("Added Algebra domain: " + i2s(b2i(add_algebra)));

let add_typefn = addDomainToNode(g, expr_id, typefn_id);
println("Added Type(\"number\") domain: " + i2s(b2i(add_typefn)));

let add_nested = addDomainToNode(g, expr_id, nested_id);
println("Added Group(a + b) domain: " + i2s(b2i(add_nested)));

// Test pattern matching with domains
println("\nTesting simple domain matching");
let matches_algebra = matchOGraphPattern(g, quote((x + y) : Algebra), \(bindings : ast, eclassId) . (
    println("MATCH FOUND with Algebra at ID: " + i2s(eclassId));
    println("Bindings: " + prettyOrbit(bindings));
));
println("Total matches with Algebra: " + i2s(matches_algebra));

// Fix the syntax for function-based domain patterns
println("\nTesting complex function domain matching");
let matches_typefn = matchOGraphPattern(g, quote((x + y) : Type("number")), \(bindings : ast, eclassId) . (
    println("MATCH FOUND with Type(\"number\") at ID: " + i2s(eclassId));
    println("Bindings: " + prettyOrbit(bindings));
));
println("Total matches with Type(\"number\"): " + i2s(matches_typefn));

println("\nTesting nested domain matching");
let matches_nested = matchOGraphPattern(g, quote((x + y) : Group(a + b)), \(bindings : ast, eclassId) . (
    println("MATCH FOUND with Group(a + b) at ID: " + i2s(eclassId));
    println("Bindings: " + prettyOrbit(bindings));
));
println("Total matches with Group(a + b): " + i2s(matches_nested));

// Test negative domain matching
println("\nTesting negative domain matching");
let non_domain = addOGraph(g, quote(NonExistentDomain));
let matches_negative = matchOGraphPattern(g, quote((x + y) !: NonExistentDomain), \(bindings : ast, eclassId) . (
    println("NEGATIVE MATCH FOUND at ID: " + i2s(eclassId));
    println("Bindings: " + prettyOrbit(bindings));
));
println("Total negative matches: " + i2s(matches_negative));

// --- NEW TESTS: findOGraphId ---
println("\nTesting findOGraphId on inserted and new terms");

let t1_id = addOGraph(g, quote(foo(bar, baz)));
println("Inserted foo(bar, baz) with ID: " + i2s(t1_id));

let found1 = findOGraphId(g, quote(foo(bar, baz)));
println("findOGraphId(foo(bar, baz)) returns: " + i2s(found1));

let t2 = quote(foo(bar, 7));
let t2_id = addOGraph(g, t2);
let found2 = findOGraphId(g, t2);
println("Inserted foo(bar, 7) with ID: " + i2s(t2_id));
println("findOGraphId(foo(bar, 7)) returns: " + i2s(found2));

let not_found = findOGraphId(g, quote(foo(bar, 999)));
println("findOGraphId(foo(bar, 999)) returns (should be -1): " + i2s(not_found));

// Also test for \"Type(\"number\")\" domain structure, which used to fail
let foundTypeFn = findOGraphId(g, quote(Type("number")));
println("findOGraphId(Type(\"number\")) returns: " + i2s(foundTypeFn) + " (should match typefn_id: " + i2s(typefn_id) + ")");

// Test for structurally-equal, canonicalized node (a+b) vs inserted expr_id
let foundExpr = findOGraphId(g, quote(a + b));
println("findOGraphId(a + b) returns: " + i2s(foundExpr) + " (should match expr_id: " + i2s(expr_id) + ")");

"Complex domain pattern test completed"
