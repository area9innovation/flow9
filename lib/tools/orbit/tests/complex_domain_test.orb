// Test of domain pattern matching with complex domain expressions

// Define quote
fn quote(e : ast) = e;

// Create a test graph
let g = makeOGraph("complex_domain_test");

// Create domain expressions and a value to annotate
let a_id = addOGraph(g, quote(a));
let b_id = addOGraph(g, quote(b));
let expr_id = addOGraph(g, quote(a + b));

// Add a simple domain
let algebra_id = addOGraph(g, quote(Algebra));

// Add a complex domain (function call)
let typefn_id = addOGraph(g, quote(Type("number")));

// Add a nested domain (a + b could be an expression or a domain)
let nested_id = addOGraph(g, quote(Group(a + b)));

// Print info about initial graph state
println("Created test graph with nodes:");
println("a with ID: " + i2s(a_id));
println("b with ID: " + i2s(b_id));
println("a + b with ID: " + i2s(expr_id));
println("Algebra with ID: " + i2s(algebra_id));
println("Type(\"number\") with ID: " + i2s(typefn_id));
println("Group(a + b) with ID: " + i2s(nested_id));

// Add domains to the expression
println("\nAdding domains to a + b");
let add_algebra = addDomainToNode(g, expr_id, algebra_id);
println("Added Algebra domain: " + i2s(b2i(add_algebra)));

let add_typefn = addDomainToNode(g, expr_id, typefn_id);
println("Added Type(\"number\") domain: " + i2s(b2i(add_typefn)));

let add_nested = addDomainToNode(g, expr_id, nested_id);
println("Added Group(a + b) domain: " + i2s(b2i(add_nested)));

// Test pattern matching with domains
println("\nTesting simple domain matching");
let matches_algebra = matchOGraphPattern(g, quote((x + y) : Algebra), \(bindings : ast, eclassId) . (
    println("MATCH FOUND with Algebra at ID: " + i2s(eclassId));
    println("Bindings: " + prettyOrbit(bindings));
));
println("Total matches with Algebra: " + i2s(matches_algebra));

// Fix the syntax for function-based domain patterns
println("\nTesting complex function domain matching");
let matches_typefn = matchOGraphPattern(g, quote((x + y) : Type("number")), \(bindings : ast, eclassId) . (
    println("MATCH FOUND with Type(\"number\") at ID: " + i2s(eclassId));
    println("Bindings: " + prettyOrbit(bindings));
));
println("Total matches with Type(\"number\"): " + i2s(matches_typefn));

println("\nTesting nested domain matching");
let matches_nested = matchOGraphPattern(g, quote((x + y) : Group(a + b)), \(bindings : ast, eclassId) . (
    println("MATCH FOUND with Group(a + b) at ID: " + i2s(eclassId));
    println("Bindings: " + prettyOrbit(bindings));
));
println("Total matches with Group(a + b): " + i2s(matches_nested));

// Test negative domain matching
println("\nTesting negative domain matching");
let non_domain = addOGraph(g, quote(NonExistentDomain));
let matches_negative = matchOGraphPattern(g, quote((x + y) !: NonExistentDomain), \(bindings : ast, eclassId) . (
    println("NEGATIVE MATCH FOUND at ID: " + i2s(eclassId));
    println("Bindings: " + prettyOrbit(bindings));
));
println("Total negative matches: " + i2s(matches_negative));

"Complex domain pattern test completed"