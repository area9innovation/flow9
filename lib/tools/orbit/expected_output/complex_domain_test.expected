Created test graph with nodes:
a with ID: 1
b with ID: 2
a + b with ID: 6
Algebra with ID: 7
Type("number") with ID: 11
Group(a + b) with ID: 18
Adding domains to a + b
Added Algebra domain: 1
Added Type("number") domain: 1
Added Group(a + b) domain: 1
Testing simple domain matching
MATCH FOUND with Algebra at ID: 6
Bindings: [Pair("x", 4), Pair("y", 5)]
Total matches with Algebra: 1
Testing complex function domain matching
	TODO: We expect a hit here
Total matches with Type("number"): 0
Testing nested domain matching
	TODO: We expect a hit here
Total matches with Group(a + b): 0
Testing negative domain matching
NEGATIVE MATCH FOUND at ID: 6
Bindings: [Pair("x", 4), Pair("y", 5)]
NEGATIVE MATCH FOUND at ID: 17
Bindings: [Pair("x", 15), Pair("y", 16)]
Total negative matches: 2
Result: "Complex domain pattern test completed"
